README.md
# progress_dialog

A light weight package to show progress dialog. As it is a stateful widget, you can change the text shown on the dialog dynamically.

[![](https://img.shields.io/badge/dontate-Buy%20Me%20a%20Coffee-blueviolet)](https://www.buymeacoffee.com/fayaz) &nbsp; 
[![LinkedIn](https://img.shields.io/badge/LinkedIn-in-0e76a8)](https://www.linkedin.com/in/fayaz07) &nbsp; [![Follow](https://img.shields.io/github/followers/fayaz07?style=social)](https://github.com/fayaz07) &nbsp; [![Fork](https://img.shields.io/github/forks/fayaz07/progress_dialog?style=social)](https://github.com/fayaz07/progress_dialog/fork) &nbsp; [![Star](https://img.shields.io/github/stars/fayaz07/progress_dialog?style=social)](https://github.com/fayaz07/progress_dialog/star) &nbsp; [![Watches](https://img.shields.io/github/watchers/fayaz07/progress_dialog?style=social)](https://github.com/fayaz07/progress_dialog/) 

[![Get the library](https://img.shields.io/badge/Get%20library-pub-blue)](https://pub.dev/packages/progress_dialog) &nbsp; [![Example](https://img.shields.io/badge/Example-Ex-success)](https://pub.dev/packages/progress_dialog#-example-tab-)



## Supported Dart Versions
**Dart SDK version >= 2.7.0**


## Demo

<img src="https://raw.githubusercontent.com/fayaz07/progress_dialog/master/progress_dialog_1.png" height="35%" width="35%"  alt="Normal dialog Demo"/> <img src="https://raw.githubusercontent.com/fayaz07/progress_dialog/master/progress_dialog_2.png"   height="35%" width="35%" alt="Download dialog Demo" />


## Installation
[![Pub](https://img.shields.io/badge/pub-1.2.0-blue)](https://pub.dev/packages/progress_dialog)

Add the Package
```yaml
dependencies:
  progress_dialog: ^1.2.4
```

## How to use



Import the package in your dart file

```dart
import 'package:progress_dialog/progress_dialog.dart';
```
Create and initialise a **ProgressDialog** object inside the **build()** method passing context to it

<ol>
<li> Initialize the ProgressDialog object <br>  

```dart
final ProgressDialog pr = ProgressDialog(context);
```
</li>


<li> By default it is a normal dialog to show some message, if you would like to use it to show percentage of progress done, specify the optional <b>type</b> parameter and specify if you want your dialog to dismiss when back button is pressed <b>isDismissible</b> parameter (Optional) <br/>

```dart
//For normal dialog
pr = ProgressDialog(context,type: ProgressDialogType.Normal, isDismissible: true/false, showLogs: true/false);
    
//For showing progress percentage
pr =  ProgressDialog(context,type: ProgressDialogType.Download, isDismissible: true/false, showLogs: true/false);
```
</li>
> Note: Please initialize the ```ProgressDialog```, where you have availability of the context
  
<li>Style the progress dialog (Optional)<br/>

```dart
pr.style(
  message: 'Downloading file...',
  borderRadius: 10.0,
  backgroundColor: Colors.white,
  progressWidget: CircularProgressIndicator(),
  elevation: 10.0,
  insetAnimCurve: Curves.easeInOut,
  progress: 0.0,
  textDirection: TextDirection.rtl,
  maxProgress: 100.0,
  progressTextStyle: TextStyle(
     color: Colors.black, fontSize: 13.0, fontWeight: FontWeight.w400),
  messageTextStyle: TextStyle(
     color: Colors.black, fontSize: 19.0, fontWeight: FontWeight.w600)
  );
```
```note
Note: You don't need to use all parameters, all of them are optional
```

</li>

<li>Showing the progress dialog<br>
  
```dart
await pr.show();
```
</li>

<li>
Dynamically update the content shown out there <br/>
  
```dart
pr.update(
  progress: 50.0,
  message: "Please wait...",
  progressWidget: Container(
    padding: EdgeInsets.all(8.0), child: CircularProgressIndicator()),
  maxProgress: 100.0,
  progressTextStyle: TextStyle(
    color: Colors.black, fontSize: 13.0, fontWeight: FontWeight.w400),
  messageTextStyle: TextStyle(
    color: Colors.black, fontSize: 19.0, fontWeight: FontWeight.w600),
  );
```
```note
Note: You don't need to use all parameters, all of them are optional
```
</li>


<li>
Dismissing the progress dialog<br/>
  
```dart
pr.hide().then((isHidden) {
  print(isHidden);
});

// or
await pr.hide();
```

</li>

</ol>

> Navigating to next screens must be done after the completion of Future - hide(). [See here for example](https://gist.github.com/fayaz07/52dfe04d9f8594501465f169aa6ef6b6)

### Check if progress dialog is showing

```dart
bool isProgressDialogShowing = pr.isShowing();
print(isProgressDialogShowing);
```

### Use custom body
```dart
    pr = ProgressDialog(
      context,
      type: ProgressDialogType.Normal,
      isDismissible: true,
      /// your body here
      customBody: LinearProgressIndicator(
        valueColor: AlwaysStoppedAnimation<Color>(Colors.blueAccent),
        backgroundColor: Colors.white,
      ),
    );
```


---


## Demo

<img src="https://raw.githubusercontent.com/fayaz07/progress_dialog/master/pd_demo_nd.gif" alt="Normal dialog Demo"/> <img src="https://raw.githubusercontent.com/fayaz07/progress_dialog/master/pd_demo_dd.gif" alt="Download dialog Demo" />


### Default configuration/styles

If you don't like to configure/style the dialog and continue with the default style, it's okay but just have a look at our default configuration.

| Attribute           | Value   |
| -------------       | :-----:|
| Dismissible         |    true |
| ProgressDialogType  |    ProgressDialogType.Normal |
| BackgroundColor     | Colors.white |
| BorderRadius        |   RoundedRectangularBorder(radius: 8.0) |
| AnimationCurve      |    Curves.easeInOut |
| Elevation           |    8.0 |
| ProgressWidget      |    Double_rings_loding_indicator |
| MessageTextStyle    |    color: Colors.black, fontSize: 19.0, fontWeight: FontWeight.w600 |
| ProgressTextStyle   |    color: Colors.black, fontSize: 13.0, fontWeight: FontWeight.w400 |
| showLogs            |    false |

#### Well let's discuss limits for configuring it

| Attribute           | Can be updated during instantiating   | Can be updated during styling   |Can be updated during dialog is shown |
| -------------       | :-------------------------------:     |:------------------------------: |:-------------------------------:     |
| Dismissible         | Yes                                   |No                               |No                                    |
| ProgressDialogType  | Yes                                   |No                               |No                                    |
| BackgroundColor     | No                                    |Yes                              |No                                    |
| BorderRadius        | No                                    |Yes                              |No                                    |
| AnimationCurve      | No                                    |Yes                              |No                                    |
| Elevation           | No                                    |Yes                              |No                                    |
| ProgressWidget      | No                                    |Yes                              |Yes                                   |
| MessageTextStyle    | No                                    |Yes                              |Yes                                   |
| ProgressTextStyle   | No                                    |Yes                              |Yes                                   |
| ShowLogs            | Yes                                   |No                               |No                                    |

---

### Want to contribute? 
Pull requests and issues are always welcome!

#### How to contribute?
<ol>
  <li> Fork the <a href="https://github.com/fayaz07/progress_dialog">repository</a></li>
  <li> Clone it to your local machine </li>
  <li> Open the project in your favourite editor </li>  
  <li> Open cmd/terminal and run <b>flutter clean</b> and then <b>flutter packages get</b> </li>
  <li> Make the changes </li>
  <li> Create a <b>Pull Request</b> </li>
</ol>

#### View the issues [here](https://github.com/fayaz07/progress_dialog/issues)

#### This library is only tested for Android, iOS contributors are most welcome

---
Loading indicator -> https://loading.io/


CHANGELOG.md
## 1.2.4
* Fixed issue RenderBox overflow
* Added rtl and ltr support

## 1.2.3
* Fixed issue FlutterError is not a subtype of string

## 1.2.2
* Added param for aligning text message
* Added custom body feature

## 1.2.1
* Added **await** for showing the dialog
* Added dialog show time delay
## 1.2.0  

 - Added custom styling for widget in runtime, background color,
   elevation, border-radius, text-styles, anim-curve, max-progress,
   no-drop on backbutton press and fixed old bugs

## 1.1.0+1
* Added interactive double ring loading indicator, updated documentation

## 1.0.8
* Added interactive double ring loading indicator, updated documentation

## 1.0.4 
 * Bug fixes.

## 1.0.3 
* Update Dart Version.

## 1.0.2
* Added shown/hidden accessibility.

## 1.0.1
* Added example and ready for package release.



example/lib/ppd.dart
import 'package:flutter/material.dart';
import 'package:flutter/cupertino.dart';

String _dialogMessage = "Loading...";
enum ProgressDialogType { Normal, Download }

ProgressDialogType _progressDialogType = ProgressDialogType.Normal;
double _progress = 0.0;

class ProgressDialog {
  _MyDialog _dialog;
  bool _isShowing = false;
  BuildContext _buildContext, _context;

  ProgressDialog(
      BuildContext buildContext, ProgressDialogType progressDialogtype) {
    _buildContext = buildContext;

    _progressDialogType = progressDialogtype;
  }

  void setMessage(String mess) {
    _dialogMessage = mess;
  }

  void update({double progress,String message}) {
    if(_progressDialogType == ProgressDialogType.Download)
      _progress = progress;
    _dialogMessage = message;
    _dialog.update();
  }

  bool isShowing() {
    return _isShowing;
  }

  void hide() {
    _isShowing = false;
    _dialogMessage = "Loading...";
    _progress = 0.0;
    Navigator.of(_context).pop();
  }

  void show() {
    _dialog = new _MyDialog();
    showDialog<dynamic>(
      context: _buildContext,
      barrierDismissible: false,
      builder: (BuildContext context) {
        _context = context;
        return Dialog(
            insetAnimationCurve: Curves.easeInOut,
            insetAnimationDuration: Duration(milliseconds: 100),
            elevation: 10.0,
            shape: RoundedRectangleBorder(borderRadius: BorderRadius.all(Radius.circular(10.0))),
            child: _dialog);
      },
    );
  }
}

class _MyDialog extends StatefulWidget {
  var _dialog = new _MyDialogState();

  update() {
    _dialog.setState(() {});
  }

  @override
  State<StatefulWidget> createState() {
    return _dialog;
  }
}

class _MyDialogState extends State<_MyDialog> {
  @override
  Widget build(BuildContext context) {
    return SizedBox(
        height: 100.0,
        child: Row(children: <Widget>[
          const SizedBox(width: 15.0),
          SizedBox(
            width: 60.0,
            child: Image.asset('assets/double_ring_loading_io.gif'),
          ),
          const SizedBox(width: 15.0),
          Expanded(
            child: _progressDialogType == ProgressDialogType.Normal
                ? Text(_dialogMessage,
                    textAlign: TextAlign.justify,
                    style: TextStyle(color: Colors.black, fontSize: 22.0,fontWeight: FontWeight.w700))
                : Stack(
                    children: <Widget>[
                      Positioned(
                        child: Text(_dialogMessage,
                            style: TextStyle(color: Colors.black, fontSize: 22.0,fontWeight: FontWeight.w700)),
                        top: 35.0,
                      ),
                      Positioned(
                        child: Text("$_progress/100",
                            style: TextStyle(color: Colors.black, fontSize: 15.0,fontWeight: FontWeight.w400)),
                        bottom: 15.0,
                        right: 15.0,
                      ),

                    ],
                  ),
          )
        ]));
  }
}

class MessageBox {
  bool _isShowing = false;

  BuildContext buildContext;
  String message = " ", title = " ";

  MessageBox(this.buildContext, this.message, this.title);

  void show() {
    _showDialog();
  }

  Future _showDialog() {
    showDialog(
      context: buildContext,
      barrierDismissible: false,
      builder: (BuildContext context) {
        return CupertinoAlertDialog(
          title: Text('$title'),
          actions: <Widget>[
            FlatButton(
              child: Text('Ok'),
              onPressed: () {
                Navigator.of(context).pop();
              },
            )
          ],
          content: SizedBox(
            height: 45.0,
            child: Center(
              child: Row(
                children: <Widget>[
                  SizedBox(width: 10.0),
                  Expanded(
                    child: Text(
                      message,
                      textAlign: TextAlign.center,
                      style: TextStyle(color: Colors.black, fontSize: 18.0),
                    ),
                  ),
                  SizedBox(width: 10.0),
                ],
              ),
            ),
          ),
        );
      },
    );
    return null;
  }
}


lib/progress_dialog.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/painting.dart';

enum ProgressDialogType { Normal, Download }

String _dialogMessage = "Loading...";
double _progress = 0.0, _maxProgress = 100.0;

Widget _customBody;

TextAlign _textAlign = TextAlign.left;
Alignment _progressWidgetAlignment = Alignment.centerLeft;

TextDirection _direction = TextDirection.ltr;

bool _isShowing = false;
BuildContext _context, _dismissingContext;
ProgressDialogType _progressDialogType;
bool _barrierDismissible = true, _showLogs = false;

TextStyle _progressTextStyle = TextStyle(
        color: Colors.black, fontSize: 12.0, fontWeight: FontWeight.w400),
    _messageStyle = TextStyle(
        color: Colors.black, fontSize: 18.0, fontWeight: FontWeight.w600);

double _dialogElevation = 8.0, _borderRadius = 8.0;
Color _backgroundColor = Colors.white;
Curve _insetAnimCurve = Curves.easeInOut;
EdgeInsets _dialogPadding = const EdgeInsets.all(8.0);

Widget _progressWidget = Image.asset(
  'assets/double_ring_loading_io.gif',
  package: 'progress_dialog',
);

class ProgressDialog {
  _Body _dialog;

  ProgressDialog(BuildContext context,
      {ProgressDialogType type,
        bool isDismissible,
        bool showLogs,
        TextDirection textDirection,
        Widget customBody}) {
    _context = context;
    _progressDialogType = type ?? ProgressDialogType.Normal;
    _barrierDismissible = isDismissible ?? true;
    _showLogs = showLogs ?? false;
    _customBody = customBody ?? null;
    _direction = textDirection ?? TextDirection.ltr;
  }

  void style(
      {Widget child,
      double progress,
      double maxProgress,
      String message,
      Widget progressWidget,
      Color backgroundColor,
      TextStyle progressTextStyle,
      TextStyle messageTextStyle,
      double elevation,
      TextAlign textAlign,
      double borderRadius,
      Curve insetAnimCurve,
      EdgeInsets padding,
      Alignment progressWidgetAlignment}) {
    if (_isShowing) return;
    if (_progressDialogType == ProgressDialogType.Download) {
      _progress = progress ?? _progress;
    }

    _dialogMessage = message ?? _dialogMessage;
    _maxProgress = maxProgress ?? _maxProgress;
    _progressWidget = progressWidget ?? _progressWidget;
    _backgroundColor = backgroundColor ?? _backgroundColor;
    _messageStyle = messageTextStyle ?? _messageStyle;
    _progressTextStyle = progressTextStyle ?? _progressTextStyle;
    _dialogElevation = elevation ?? _dialogElevation;
    _borderRadius = borderRadius ?? _borderRadius;
    _insetAnimCurve = insetAnimCurve ?? _insetAnimCurve;
    _textAlign = textAlign ?? _textAlign;
    _progressWidget = child ?? _progressWidget;
    _dialogPadding = padding ?? _dialogPadding;
    _progressWidgetAlignment =
        progressWidgetAlignment ?? _progressWidgetAlignment;
  }

  void update(
      {double progress,
      double maxProgress,
      String message,
      Widget progressWidget,
      TextStyle progressTextStyle,
      TextStyle messageTextStyle}) {
    if (_progressDialogType == ProgressDialogType.Download) {
      _progress = progress ?? _progress;
    }

    _dialogMessage = message ?? _dialogMessage;
    _maxProgress = maxProgress ?? _maxProgress;
    _progressWidget = progressWidget ?? _progressWidget;
    _messageStyle = messageTextStyle ?? _messageStyle;
    _progressTextStyle = progressTextStyle ?? _progressTextStyle;

    if (_isShowing) _dialog.update();
  }

  bool isShowing() {
    return _isShowing;
  }

  Future<bool> hide() async {
    try {
      if (_isShowing) {
        _isShowing = false;
        Navigator.of(_dismissingContext).pop();
        if (_showLogs) debugPrint('ProgressDialog dismissed');
        return Future.value(true);
      } else {
        if (_showLogs) debugPrint('ProgressDialog already dismissed');
        return Future.value(false);
      }
    } catch (err) {
      debugPrint('Seems there is an issue hiding dialog');
      debugPrint(err.toString());
      return Future.value(false);
    }
  }

  Future<bool> show() async {
    try {
      if (!_isShowing) {
        _dialog = new _Body();
        showDialog<dynamic>(
          context: _context,
          barrierDismissible: _barrierDismissible,
          builder: (BuildContext context) {
            _dismissingContext = context;
            return WillPopScope(
              onWillPop: () async => _barrierDismissible,
              child: Dialog(
                  backgroundColor: _backgroundColor,
                  insetAnimationCurve: _insetAnimCurve,
                  insetAnimationDuration: Duration(milliseconds: 100),
                  elevation: _dialogElevation,
                  shape: RoundedRectangleBorder(
                      borderRadius:
                          BorderRadius.all(Radius.circular(_borderRadius))),
                  child: _dialog),
            );
          },
        );
        // Delaying the function for 200 milliseconds
        // [Default transitionDuration of DialogRoute]
        await Future.delayed(Duration(milliseconds: 200));
        if (_showLogs) debugPrint('ProgressDialog shown');
        _isShowing = true;
        return true;
      } else {
        if (_showLogs) debugPrint("ProgressDialog already shown/showing");
        return false;
      }
    } catch (err) {
      _isShowing = false;
      debugPrint('Exception while showing the dialog');
      debugPrint(err.toString());
      return false;
    }
  }
}

// ignore: must_be_immutable
class _Body extends StatefulWidget {
  _BodyState _dialog = _BodyState();

  update() {
    _dialog.update();
  }

  @override
  State<StatefulWidget> createState() {
    return _dialog;
  }
}

class _BodyState extends State<_Body> {
  update() {
    setState(() {});
  }

  @override
  void dispose() {
    _isShowing = false;
    if (_showLogs) debugPrint('ProgressDialog dismissed by back button');
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    final loader = Align(
      alignment: _progressWidgetAlignment,
      child: SizedBox(
        width: 60.0,
        height: 60.0,
        child: _progressWidget,
      ),
    );

    final text = Expanded(
      child: _progressDialogType == ProgressDialogType.Normal
          ? Text(
        _dialogMessage,
        textAlign: _textAlign,
        style: _messageStyle,
        textDirection: _direction,
      )
          : Padding(
        padding: const EdgeInsets.all(8.0),
        child: Column(
          mainAxisSize: MainAxisSize.min,
          children: <Widget>[
            SizedBox(height: 8.0),
            Row(
              children: <Widget>[
                Expanded(
                    child: Text(
                      _dialogMessage,
                      style: _messageStyle,
                      textDirection: _direction,
                    )),
              ],
            ),
            SizedBox(height: 4.0),
            Align(
              alignment: Alignment.bottomRight,
              child: Text(
                "$_progress/$_maxProgress",
                style: _progressTextStyle,
                textDirection: _direction,
              ),
            ),
          ],
        ),
      ),
    );

    return _customBody ??
        Container(
          padding: _dialogPadding,
          child: Column(
            mainAxisSize: MainAxisSize.min,
            children: <Widget>[
              // row body
              Row(
                mainAxisSize: MainAxisSize.min,
                children: <Widget>[
                  const SizedBox(width: 8.0),
                  _direction == TextDirection.ltr ? loader : text,
                  const SizedBox(width: 8.0),
                  _direction == TextDirection.rtl ? loader : text,
                  const SizedBox(width: 8.0)
                ],
              ),
            ],
          ),
        );
  }
}


example/lib/issues/issue_44.dart
//import 'package:flutter/cupertino.dart';
//import 'package:flutter/material.dart';
//import 'package:progress_dialog/progress_dialog.dart';
//
//void main() => runApp(MaterialApp(home: IndexPage()));
//
//class IndexPage extends StatefulWidget {
//  IndexPage({Key key}) : super(key: key);
//
//  @override
//  _IndexPageState createState() => _IndexPageState();
//}
//
//// AutomaticKeepAliveClientMixin, see https://blog.csdn.net/a875801/article/details/92829950
//// TickerProviderStateMixin, see https://blog.csdn.net/u011272795/article/details/82740389
//class _IndexPageState extends State<IndexPage>
//    with AutomaticKeepAliveClientMixin, TickerProviderStateMixin {
////  List<Course> recommendCourses = [];
////  List<Course> freeCourses = [];
////  List<News> recommendNews = [];
//  ProgressDialog pr;
//  TabController _tabController;
//
//  @override
//  bool get wantKeepAlive => true;
//
//  @override
//  void initState() {
//    print('Index, initState');
//
//    // pr = new ProgressDialog(context, showLogs: true);
////    _tabController = TabController(length: categories.length, vsync: this);
//    _tabController = TabController(length: 3, vsync: this);
//    _updateState();
//    super.initState();
//  }
//
//  @override
//  void dispose() {
//    print('Index, dispose');
//    _tabController.dispose();
//    super.dispose();
//  }
//
//  void _updateState() async {
//    pr.show();
//    print('pr.show()');
//    var _recommendCourses = await CommonService().getRecommendCourses(3);
//    var _freeCourses = await CommonService().getfreeCourses(2);
//    var _recommendNews = await CommonService().getRecommendNews(3);
//    print('_updateState: ${_recommendCourses}');
//    setState(() {
//      print('setState: ${_recommendCourses}');
//      recommendCourses = _recommendCourses;
//      freeCourses = _freeCourses;
//      recommendNews = _recommendNews;
//      pr.hide();
//      print('pr.hide()');
//    });
//  }
////...
//}


example/lib/issues/issue_57.dart
import 'package:flutter/material.dart';
import 'package:progress_dialog/progress_dialog.dart';
import 'dart:async';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  Widget build(context) {
    return MaterialApp(
      title: 'PDTest',
      initialRoute: '/home', 
      routes: {
        '/home': (context) {
          return HomePage();
        },
        '/profile': (conetxt) {
          return ProfilePage();
        },
      },
    );
  }
}

class HomePage extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    ProgressDialog _pr = ProgressDialog(
      context,
      type: ProgressDialogType.Normal,
      isDismissible: false,
    );
    return Scaffold(
      appBar: AppBar(
        title: Text('Home Page'),
      ),
      drawer: drawerMenu(_pr, context),
      body: Center(
        child: Text('Home Page'),
      ),
    );
  }
}

Widget drawerMenu(ProgressDialog _pr, BuildContext context) {
  
  return Drawer(
    child: ListView(
      children: <Widget>[
        DrawerHeader(child: null),
        ListTile(
          title: Text('Profile Page'),
          onTap: () async {
            await _pr.show();
            Timer(Duration(seconds: 3), () async {
              if (_pr.isShowing()) await _pr.hide();
              Navigator.pushReplacementNamed(context, '/profile');
            });
          },
        ),
      ],
    ),
  );
}

class ProfilePage extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    ProgressDialog _pr = ProgressDialog(
      context,
      type: ProgressDialogType.Normal,
      isDismissible: false,
    );
    return Scaffold(
      appBar: AppBar(
        leading: IconButton(
          icon: Icon(Icons.arrow_back),
          onPressed: (){
            Navigator.of(context).pushReplacementNamed('/home');
          },
        ),
        title: Text('Profile Page'),
      ),
      body: Center(
        child: Text('Profile Page'),
      ),
    );
  }
}

example/lib/main.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:progress_dialog/progress_dialog.dart';

//import '../../lib/progress_dialog.dart';

ProgressDialog pr;

void main() {
  runApp(MaterialApp(
    home: MyApp(),
  ));
}

class MyApp extends StatelessWidget {
  double percentage = 0.0;

  @override
  Widget build(BuildContext context) {
//    pr = new ProgressDialog(context,
//        type: ProgressDialogType.Normal, isDismissible: false);
//    pr = new ProgressDialog(context, type: ProgressDialogType.Download);

// Custom body test
    pr = ProgressDialog(
      context,
      type: ProgressDialogType.Download,
      textDirection: TextDirection.rtl,
      isDismissible: true,
//      customBody: LinearProgressIndicator(
//        valueColor: AlwaysStoppedAnimation<Color>(Colors.blueAccent),
//        backgroundColor: Colors.white,
//      ),
    );

    pr.style(
//      message: 'Downloading file...',
      message:
          'Lets dump some huge text into the progress dialog and check whether it can handle the huge text. If it works then not you or me, flutter is awesome',
      borderRadius: 10.0,
      backgroundColor: Colors.white,
      elevation: 10.0,
      insetAnimCurve: Curves.easeInOut,
      progress: 0.0,
      progressWidgetAlignment: Alignment.center,
      maxProgress: 100.0,
      progressTextStyle: TextStyle(
          color: Colors.black, fontSize: 13.0, fontWeight: FontWeight.w400),
      messageTextStyle: TextStyle(
          color: Colors.black, fontSize: 19.0, fontWeight: FontWeight.w600),
    );

    return Scaffold(
      body: Center(
        child: RaisedButton(
            child: Text(
              'Show Dialog',
              style: TextStyle(color: Colors.white),
            ),
            color: Colors.blue,
            onPressed: () async {
              await pr.show();

              Future.delayed(Duration(seconds: 2)).then((onvalue) {
                percentage = percentage + 30.0;
                print(percentage);

                pr.update(
                  progress: percentage,
                  message: "Please wait...",
                  progressWidget: Container(
                      padding: EdgeInsets.all(8.0),
                      child: CircularProgressIndicator()),
                  maxProgress: 100.0,
                  progressTextStyle: TextStyle(
                      color: Colors.black,
                      fontSize: 13.0,
                      fontWeight: FontWeight.w400),
                  messageTextStyle: TextStyle(
                      color: Colors.black,
                      fontSize: 19.0,
                      fontWeight: FontWeight.w600),
                );

                Future.delayed(Duration(seconds: 2)).then((value) {
                  percentage = percentage + 30.0;
                  pr.update(
                      progress: percentage, message: "Few more seconds...");
                  print(percentage);
                  Future.delayed(Duration(seconds: 2)).then((value) {
                    percentage = percentage + 30.0;
                    pr.update(progress: percentage, message: "Almost done...");
                    print(percentage);

                    Future.delayed(Duration(seconds: 2)).then((value) {
                      pr.hide().whenComplete(() {
                        print(pr.isShowing());
                      });
                      percentage = 0.0;
                    });
                  });
                });
              });

              Future.delayed(Duration(seconds: 10)).then((onValue) {
                print("PR status  ${pr.isShowing()}");
                if (pr.isShowing())
                  pr.hide().then((isHidden) {
                    print(isHidden);
                  });
                print("PR status  ${pr.isShowing()}");
              });
            }),
      ),
    );
  }
}

class FirstScreen extends StatefulWidget {
  @override
  _FirstScreenState createState() => _FirstScreenState();
}

class _FirstScreenState extends State<FirstScreen> {
  ProgressDialog pr;

  @override
  Widget build(BuildContext context) {
    pr = new ProgressDialog(context, showLogs: true);
    pr.style(message: 'Please wait...');

    return Scaffold(
      body: Center(
        child: RaisedButton(
          child: Text('Show dialog and go to next screen',
              style: TextStyle(color: Colors.white)),
          color: Colors.blueAccent,
          onPressed: () {
            pr.show();
            Future.delayed(Duration(seconds: 3)).then((value) {
              pr.hide().whenComplete(() {
                Navigator.of(context).push(CupertinoPageRoute(
                    builder: (BuildContext context) => SecondScreen()));
              });
            });
          },
        ),
      ),
    );
  }
}

class SecondScreen extends StatefulWidget {
  @override
  _SecondScreenState createState() => _SecondScreenState();
}

class _SecondScreenState extends State<SecondScreen> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(child: Text('I am second screen')),
    );
  }
}
